src/Constraint/JsonValueMatchesMany.php

```diff
@@ -59,11 +62,23 @@ function (Constraint $constraint) {
      */
     protected function matches($other): bool
     {
+        $result = true;
         foreach ($this->constraints as $constraint) {
             if (!$constraint->evaluate($other, '', true)) {
-                return false;
+                $result = false;
+                $this->failedConstraints[] = $constraint->toString();
```

Comment:
This introduces a state to the class that will persist across multiple invocations. This is most probably an edge case, but _will_ lead to odd error messages when the same instance of a `JsonValueMatchesMany` constraint is re-used multiple times across a test case.

When looking at other implementations of `additionalFailureDescription` (PHPUnit itself uses it only once, in the `StringMatchesFormatDescription` constraint), these are stateless (i.e. re-compute the failure description there and then based on the `$other` input).

https://github.com/sebastianbergmann/phpunit/blob/99f6e264b464b3a9089ea2dceed79c3d953244e8/src/Framework/Constraint/String/StringMatchesFormatDescription.php#L58-L77

Reply:
Thanks for the review.

I didn't look at other implementations before attempting to work on this.
Next time, I will make sure to explore further before jumping in. 💯

I think it is better to keep the class stateless.
Stateless class is simple and easy to reason about and also more resilient to changes.

I will update the PR shortly. 👍

Reply:
I have updated the `additionalFailureDescription` method to build message statelessly.

Please kindly review again.